File: ThirdPartyAdapters/vungle/vungle/src/main/java/com/google/ads/mediation/vungle/VungleInitializer.java
package com.google.ads.mediation.vungle;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.google.ads.mediation.vungle.util.ErrorUtil;
import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.RequestConfiguration;
import com.vungle.ads.InitializationListener;
import com.vungle.ads.Plugin;
import com.vungle.ads.VungleAds;
import com.vungle.ads.VungleException;
import com.vungle.ads.VungleSettings;
import com.vungle.ads.internal.network.VungleApiClient;
import com.vungle.mediation.VungleNetworkSettings;

import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicBoolean;

public class VungleInitializer implements InitializationListener {

    private static final VungleInitializer instance = new VungleInitializer();
    private final AtomicBoolean isInitializing = new AtomicBoolean(false);
    private final ArrayList<VungleInitializationListener> initListener;
    private final Handler mHandler = new Handler(Looper.getMainLooper());

    public static VungleInitializer getInstance() {
        return instance;
    }

    private VungleInitializer() {
        Plugin.addWrapperInfo(
                VungleApiClient.WrapperFramework.admob,
                com.vungle.mediation.BuildConfig.ADAPTER_VERSION.replace('.', '_')
        );
        initListener = new ArrayList<>();
    }

    public void initialize(
            final String appId, final Context context, VungleInitializationListener listener) {

        if (VungleAds.isInitialized()) {
            listener.onInitializeSuccess();
            return;
        }

        if (isInitializing.getAndSet(true)) {
            initListener.add(listener);
            return;
        }
        VungleNetworkSettings.setVungleSettingsChangedListener(vungleSettings -> {
            if (!VungleAds.isInitialized()) {
                return;
            }

            // Pass new settings to SDK.
            updateCoppaStatus(MobileAds.getRequestConfiguration().getTagForChildDirectedTreatment());
            VungleAds.init(context.getApplicationContext(), appId, VungleInitializer.this, vungleSettings);
        });

        updateCoppaStatus(MobileAds.getRequestConfiguration().getTagForChildDirectedTreatment());

        VungleSettings vungleSettings = VungleNetworkSettings.getVungleSettings();
        VungleAds.init(context, appId, VungleInitializer.this, vungleSettings);
        initListener.add(listener);
    }


    public void updateCoppaStatus(int configuration) {
        switch (configuration) {
            case RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_TRUE:
                VungleAds.updateUserCoppaStatus(true);
                break;
            case RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_FALSE:
                VungleAds.updateUserCoppaStatus(false);
                break;
            case RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_UNSPECIFIED:
            default:
                // Vungle's SDK only supports updating a user's COPPA status with true and false
                // values. If you haven't specified how you would like your content treated with
                // respect to COPPA in ad requests, you must indicate in the Vungle Publisher
                // Dashboard whether your app is directed toward children under age 13.
                break;
        }
    }

    @Override
    public void onError(@NonNull VungleException e) {
        final AdError error = ErrorUtil.getAdError(e);
        mHandler.post(() -> {
            for (VungleInitializationListener listener : initListener) {
                listener.onInitializeError(error);
            }
            initListener.clear();
        });
        isInitializing.set(false);
    }

    @Override
    public void onSuccess() {
        mHandler.post(() -> {
            for (VungleInitializationListener listener : initListener) {
                listener.onInitializeSuccess();
            }
            initListener.clear();
        });
        isInitializing.set(false);
    }

    public interface VungleInitializationListener {

        void onInitializeSuccess();

        void onInitializeError(AdError error);

    }
}

File: ThirdPartyAdapters/vungle/vungle/src/main/java/com/google/ads/mediation/vungle/VungleInitializer.kt
package com.google.ads.mediation.vungle

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.google.ads.mediation.vungle.VungleMediationAdapter.Companion.getAdError
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.MobileAds
import com.vungle.ads.InitializationListener
import com.vungle.ads.VungleAds
import com.vungle.ads.VungleException
import com.vungle.ads.VungleSettings
import com.vungle.ads.internal.network.Plugin
import com.vungle.ads.internal.network.VungleApiClient
import com.vungle.mediation.BuildConfig
import com.vungle.mediation.VungleNetworkSettings.VungleSettingsChangedListener
import com.vungle.mediation.VungleNetworkSettings.getVungleSettings
import com.vungle.mediation.VungleNetworkSettings.setVungleSettingsChangedListener
import java.util.concurrent.atomic.AtomicBoolean

object VungleInitializer : InitializationListener {

    private val mIsInitializing = AtomicBoolean(false)

    private val mInitListeners: ArrayList<VungleInitializationListener> = ArrayList()

    private val mHandler = Handler(Looper.getMainLooper())

    fun initialize(appId: String, context: Context, listener: VungleInitializationListener) {
        if (VungleAds.isInitialized()) {
            listener.onInitializeSuccess()
            return
        }
        if (mIsInitializing.getAndSet(true)) {
            mInitListeners.add(listener)
            return
        }

        // Keep monitoring VungleSettings in case of any changes we need to re-init SDK to apply
        // updated settings.
        setVungleSettingsChangedListener(
            object : VungleSettingsChangedListener {
                override fun onVungleSettingsChanged(settings: VungleSettings) {
                    // Ignore if sdk is yet to initialize, it will get considered while init.
                    if (!VungleAds.isInitialized()) {
                        return
                    }

                    // Pass new settings to SDK.
                    updateCoppaStatus(
                        MobileAds.getRequestConfiguration().tagForChildDirectedTreatment
                    )
                    VungleAds.init(context, appId, this@VungleInitializer, settings)
                }
            })
        updateCoppaStatus(MobileAds.getRequestConfiguration().tagForChildDirectedTreatment)
        val vungleSettings: VungleSettings = getVungleSettings()
        VungleAds.init(context.applicationContext, appId, this@VungleInitializer, vungleSettings)
        mInitListeners.add(listener)
    }

    override fun onSuccess() {
        mHandler.post {
            for (listener in mInitListeners) {
                listener.onInitializeSuccess()
            }
            mInitListeners.clear()
        }
        mIsInitializing.set(false)
    }

    override fun onError(vungleException: VungleException) {
        val error = getAdError(vungleException!!)
        mHandler.post {
            for (listener in mInitListeners) {
                listener.onInitializeError(error)
            }
            mInitListeners.clear()
        }
        mIsInitializing.set(false)
    }

    fun onAutoCacheAdAvailable(placementId: String?) {
        // Unused
    }

    fun updateCoppaStatus(configuration: Int) {
        // TODO: Update coppa status
//        when (configuration) {
//            RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_TRUE -> Vungle.updateUserCoppaStatus(
//                true
//            )
//            RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_FALSE -> Vungle.updateUserCoppaStatus(
//                false
//            )
//            RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_UNSPECIFIED -> {}
//            else -> {}
//        }
    }

    interface VungleInitializationListener {

        fun onInitializeSuccess()

        fun onInitializeError(error: AdError?)
    }

    init {
        Plugin.addWrapperInfo(
            VungleApiClient.WrapperFramework.admob,
            BuildConfig.ADAPTER_VERSION.replace('.', '_')
        )
    }
}